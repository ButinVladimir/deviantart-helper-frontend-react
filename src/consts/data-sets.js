export const DATA_SET_TIMESTAMP = 0;
export const DATA_SET_VIEWS = 1;
export const DATA_SET_FAVOURITES = 2;
export const DATA_SET_COMMENTS = 3;
export const DATA_SET_DOWNLOADS = 4;

export const dataSetOptions = [
  [DATA_SET_VIEWS, 'Views'],
  [DATA_SET_FAVOURITES, 'Favourites'],
  [DATA_SET_COMMENTS, 'Comments'],
  [DATA_SET_DOWNLOADS, 'Downloads'],
];
