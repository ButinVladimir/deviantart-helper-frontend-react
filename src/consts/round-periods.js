export const ROUND_PERIOD_NO_ROUND = 1;
export const ROUND_PERIOD_1_HOUR = 60 * 60 * 1000;
export const ROUND_PERIOD_4_HOURS = 4 * ROUND_PERIOD_1_HOUR;
export const ROUND_PERIOD_12_HOURS = 12 * ROUND_PERIOD_1_HOUR;
export const ROUND_PERIOD_1_DAY = 24 * ROUND_PERIOD_1_HOUR;

export const roundPeriodOptions = [
  [ROUND_PERIOD_NO_ROUND, 'No rounding'],
  [ROUND_PERIOD_1_HOUR, '1 hour'],
  [ROUND_PERIOD_4_HOURS, '4 hours'],
  [ROUND_PERIOD_12_HOURS, '12 hours'],
  [ROUND_PERIOD_1_DAY, '1 day'],
];
