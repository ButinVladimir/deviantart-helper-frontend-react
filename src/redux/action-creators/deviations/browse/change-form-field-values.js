import { DEVIATIONS_BROWSE_CHANGE_FORM_FIELD_VALUES } from '../../../actions';
import getTimestamp from '../../../../helpers/get-timestamp';

/**
 * @global
 * @description
 * Action to change form fields values on deviations browse page.
 *
 * @typedef {Object} DeviationsBrowseChangeFormFieldValuesAction
 */

/**
 * @description
 * Creates action to change form fields values on deviations browse page.
 *
 * @param {Object} values - New form fields values.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
const changeFormFieldValues = values => ({
  type: DEVIATIONS_BROWSE_CHANGE_FORM_FIELD_VALUES,
  ...values,
});

/**
 * @description
 * Creates action to change published time begin value on deviations browse page.
 *
 * @param {Date} publishedTimeBegin - Published time begin value.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
export const changePublishedTimeBeginActionCreator = publishedTimeBegin => changeFormFieldValues({
  publishedTimeBegin: getTimestamp(publishedTimeBegin),
});

/**
 * @description
 * Creates action to change published time end value on deviations browse page.
 *
 * @param {Date} publishedTimeEnd - Published time end value.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
export const changePublishedTimeEndActionCreator = publishedTimeEnd => changeFormFieldValues({
  publishedTimeEnd: getTimestamp(publishedTimeEnd),
});

/**
 * @description
 * Creates action to change sort field value on deviations browse page.
 *
 * @param {string} sortField - Sort field value.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
export const changeSortFieldActionCreator = sortField => changeFormFieldValues({
  sortField,
});

/**
 * @description
 * Creates action to change published time end value on deviations browse page.
 *
 * @param {string} sortOrder - Sort order value.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
export const changeSortOrderActionCreator = sortOrder => changeFormFieldValues({
  sortOrder: parseInt(sortOrder, 10),
});

/**
 * @description
 * Creates action to change title value on deviations browse page.
 *
 * @param {string} title - Title value.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
export const changeTitleActionCreator = title => changeFormFieldValues({
  title,
});

/**
 * @description
 * Creates action to change NSFW value on deviations browse page.
 *
 * @param {string} nsfw - NSFW value.
 * @returns {DeviationsBrowseChangeFormFieldValuesAction} Action.
 */
export const changeNsfwActionCreator = nsfw => changeFormFieldValues({
  nsfw,
});
