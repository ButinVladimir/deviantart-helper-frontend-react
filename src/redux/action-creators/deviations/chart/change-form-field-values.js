import { DEVIATIONS_CHART_CHANGE_FORM_FIELD_VALUES } from '../../../actions';

/**
 * @global
 * @description
 * Action to change form field values for deviations chart.
 *
 * @typedef {Object} DeviationsChartChangeFormFieldValuesAction
 */

/**
 * @description
 * Creates action to change form field values for deviations chart.
 *
 * @param {Object} values - New form fields values.
 * @returns {DeviationsChartChangeFormFieldValuesAction} Action.
 */
const changeFormFieldValues = values => ({
  type: DEVIATIONS_CHART_CHANGE_FORM_FIELD_VALUES,
  ...values,
});

/**
 * @description
 * Creates action to change data set value for deviations chart.
 *
 * @param {number} dataSet - Data set value.
 * @returns {DeviationsChartChangeFormFieldValuesAction} Action.
 */
export const changeDataSetActionCreator = dataSet => changeFormFieldValues({
  dataSet,
});

/**
 * @description
 * Creates action to change round period value for deviations chart.
 *
 * @param {number} roundPeriod - Data set value.
 * @returns {DeviationsChartChangeFormFieldValuesAction} Action.
 */
export const changeRoundPeriodActionCreator = roundPeriod => changeFormFieldValues({
  roundPeriod: parseInt(roundPeriod, 10),
});

/**
 * @description
 * Creates action to change show time value for deviations chart.
 *
 * @param {boolean} showTime - Show time value.
 * @returns {DeviationsChartChangeFormFieldValuesAction} Action.
 */
export const changeShowTimeActionCreator = showTime => changeFormFieldValues({
  showTime,
});
