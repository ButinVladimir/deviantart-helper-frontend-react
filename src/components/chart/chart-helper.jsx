import React from 'react';
import { DATA_SET_TIMESTAMP } from '../../consts/data-sets';

const timezoneOffset = -(new Date().getTimezoneOffset()) * 60 * 1000;

/**
 * @description
 * Formatter for chart.
 * Converts timestamp to a proper date string.
 *
 * @param {number} timestamp - The timestamp in the local zone.
 * @returns {string} Converted date.
 */
export const dateFormatter = timestamp => new Date(timestamp).toLocaleDateString();

/**
 * @description
 * Formatter for chart.
 * Converts timestamp to a proper datetime string.
 *
 * @param {number} timestamp - The timestamp in the local zone.
 * @returns {string} Converted datetime.
 */
export const dateTimeFormatter = timestamp => new Date(timestamp).toLocaleString();

/**
 * @description
 * Makes legend formatter for chart.
 * Converts deviation ID in legend to a title.
 *
 * @param {Map<string, string>} titlesMap - Map of ID to titles.
 * @returns {Function} Formatter.
 */
export const legendFormatter = titlesMap => (value, entry) => (
  <span style={{ color: entry.color }}>
    {titlesMap.get(value)}
  </span>
);

/**
 * @description
 * Makes tooltip formatter for chart.
 * Converts deviation ID in legend to a title.
 *
 * @param {Map<string, string>} titlesMap - Map of ID to titles.
 * @param {Map<string, string>} diffMap - Map of timestamps to maps of ID to differences.
 * @returns {Function} Formatter.
 */
export const tooltipFormatter = (titlesMap, diffMap) => (
  value,
  name,
  { payload: { timestamp } },
) => [
  `${value}${diffMap.get(timestamp).get(name)}`, titlesMap.get(name),
];


/**
 * @description
 * Rounds up the timestamp.
 * Calculates rounded timestamp in local time.
 *
 * @param {number} timestamp - The timestamp.
 * @param {number} roundPeriod - The round period.
 * @returns {number} Rounded timestamp.
 */
export const roundTimestamp = (timestamp, roundPeriod) => Math.floor(
  1 + (timestamp + timezoneOffset) / roundPeriod,
) * roundPeriod - timezoneOffset - 1;

/**
 * @description
 * Recharts doesn't merge data cells with same timestamp by itself.
 * It had to be done manually, so timestamps are splitted
 * into different buckets by selected period.
 * For each bucket, data with latest timestamp is selected.
 *
 * @param {Object} metadata - The metadata.
 * Has format of object map of deviation id to array of metadata.
 * @param {string} dataSet - Selected data set.
 * @param {number} roundPeriod - Selected round period.
 * @returns {Object[]} Parsed metadata.
 */
export const convertMetadata = (metadata, dataSet, roundPeriod) => {
  // Map of timestamps to objects with values.
  const flattenedMetadataMap = new Map();

  Object.entries(metadata).forEach((deviationMetadata) => {
    deviationMetadata[1].forEach((md) => {
      const timestamp = roundTimestamp(md[DATA_SET_TIMESTAMP], roundPeriod);
      let bucketValues;

      if (!flattenedMetadataMap.has(timestamp)) {
        bucketValues = {};
        flattenedMetadataMap.set(timestamp, bucketValues);
      } else {
        bucketValues = flattenedMetadataMap.get(timestamp);
      }

      bucketValues[deviationMetadata[0]] = md[dataSet];
    });
  });

  return Array
    .from(flattenedMetadataMap.entries())
    .map(flattenedMetadataEntry => ({
      timestamp: flattenedMetadataEntry[0],
      ...flattenedMetadataEntry[1],
    }));
};

/**
 * @description
 * Converts array of metadata into array of differences
 * for each timestamp and deviation.
 *
 * @param {Object[]} convertedMetadata - Parsed metadata.
 * @returns {Object[]} Array with differences. Has same format as metadata.
 */
export const calculateDifferences = (convertedMetadata) => {
  // Map of deviation ids to their previous values.
  const previousValuesMap = new Map();

  return convertedMetadata
    .map((md) => {
      const result = { timestamp: md.timestamp };

      Object.entries(md).forEach((entry) => {
        if (entry[0] === 'timestamp') {
          return;
        }

        if (previousValuesMap.has(entry[0])) {
          result[entry[0]] = entry[1] - previousValuesMap.get(entry[0]);
        }

        previousValuesMap.set(entry[0], entry[1]);
      });

      return result;
    })
    .filter(md => Object.entries(md).length > 1);
};

/**
 * @description
 * Makes difference map with differences for each timestamp and deviation.
 *
 * @param {Object[]} convertedMetadata - Parsed metadata.
 * @returns {Map<number, Map<string, string>>} The difference map.
 * Has format of timestamp to map of deviation id to diff string.
 */
export const makeDiffMap = (convertedMetadata) => {
  // The resulting map.
  const differencesMap = new Map();

  // Map of deviation ids to their previous values.
  const previousValuesMap = new Map();

  convertedMetadata.forEach((md) => {
    // Map of deviation ids to diff strings.
    const diffs = new Map();
    differencesMap.set(md.timestamp, diffs);

    Object.entries(md).forEach((entry) => {
      if (entry[0] === 'timestamp') {
        return;
      }

      if (!previousValuesMap.has(entry[0])) {
        diffs.set(entry[0], '');
      } else {
        const diff = entry[1] - previousValuesMap.get(entry[0]);
        diffs.set(
          entry[0],
          diff > 0 ? ` (+${diff})` : ` (${diff})`,
        );
      }

      previousValuesMap.set(entry[0], entry[1]);
    });
  });

  return differencesMap;
};
