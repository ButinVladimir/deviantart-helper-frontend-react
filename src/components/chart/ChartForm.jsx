import React from 'react';
import PropTypes from 'prop-types';
import {
  Field,
  Label as FormLabel,
  Control,
  Select,
  Checkbox,
} from 'react-bulma-components/lib/components/form';
import Columns from 'react-bulma-components/lib/components/columns';
import convertOptions from '../../helpers/convert-options';
import { roundPeriodOptions } from '../../consts/round-periods';

export default function ChartForm({
  dataSetTitlesMap,
  dataSet,
  roundPeriod,
  showTime,
  showDifferences,
  dataSetChangeHandler,
  roundPeriodChangeHandler,
  showTimeChangeHandler,
  showDifferencesChangeHandler,
}) {
  const dataSetOptionsElements = convertOptions(Array.from(dataSetTitlesMap.entries()));
  const roundPeriodOptionsElements = convertOptions(roundPeriodOptions);

  return (
    <Columns>
      <Columns.Column>
        <Field>
          <FormLabel>Data set</FormLabel>
          <Control>
            <Select
              id="data-set"
              name="data-set"
              value={dataSet}
              onChange={dataSetChangeHandler}
            >
              {dataSetOptionsElements}
            </Select>
          </Control>
        </Field>
      </Columns.Column>

      <Columns.Column>
        <Field>
          <FormLabel>Round period</FormLabel>
          <Control>
            <Select
              id="round-period"
              name="round-period"
              value={roundPeriod}
              onChange={roundPeriodChangeHandler}
            >
              {roundPeriodOptionsElements}
            </Select>
          </Control>
        </Field>
      </Columns.Column>

      <Columns.Column>
        <Field>
          <FormLabel>Show time</FormLabel>
          <Control>
            <Checkbox
              id="show-time"
              name="show-time"
              checked={showTime}
              onChange={showTimeChangeHandler}
            />
          </Control>
        </Field>
      </Columns.Column>

      <Columns.Column narrow>
        <Field>
          <FormLabel>Show differences</FormLabel>
          <Control>
            <Checkbox
              id="show-differences"
              name="show-differences"
              checked={showDifferences}
              onChange={showDifferencesChangeHandler}
            />
          </Control>
        </Field>
      </Columns.Column>
    </Columns>
  );
}

ChartForm.propTypes = {
  dataSetTitlesMap: PropTypes.instanceOf(Map).isRequired,
  dataSet: PropTypes.number.isRequired,
  roundPeriod: PropTypes.number.isRequired,
  showTime: PropTypes.bool.isRequired,
  showDifferences: PropTypes.bool.isRequired,
  dataSetChangeHandler: PropTypes.func.isRequired,
  roundPeriodChangeHandler: PropTypes.func.isRequired,
  showTimeChangeHandler: PropTypes.func.isRequired,
  showDifferencesChangeHandler: PropTypes.func.isRequired,
};
