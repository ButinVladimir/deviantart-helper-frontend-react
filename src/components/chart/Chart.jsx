import React from 'react';
import PropTypes from 'prop-types';
import Level from 'react-bulma-components/lib/components/level';
import {
  ResponsiveContainer,
  LineChart,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
  Line,
  Label,
  Legend,
} from 'recharts';
import ChartForm from './ChartForm';
import {
  convertMetadata,
  calculateDifferences,
  makeDiffMap,
  dateFormatter,
  dateTimeFormatter,
  tooltipFormatter,
  legendFormatter,
} from './chart-helper';
import colors from '../../consts/colors';

export default function Chart({
  dataSetTitlesMap,
  metadata,
  titlesMap,
  dataSet,
  roundPeriod,
  showTime,
  showDifferences,
  dataSetChangeHandler,
  roundPeriodChangeHandler,
  showTimeChangeHandler,
  showDifferencesChangeHandler,
}) {
  let convertedMetadata = convertMetadata(metadata, dataSet, roundPeriod);
  if (showDifferences) {
    convertedMetadata = calculateDifferences(convertedMetadata);
  }
  const differencesMap = makeDiffMap(convertedMetadata);

  const content = Array.from(titlesMap.keys()).map((key, index) => (
    <Line
      key={key}
      connectNulls
      type="monotone"
      dataKey={key}
      stroke={colors[index % colors.length]}
      strokeWidth={2}
    />
  ));

  return (
    <>
      <ChartForm
        dataSetTitlesMap={dataSetTitlesMap}
        dataSet={dataSet}
        roundPeriod={roundPeriod}
        showTime={showTime}
        showDifferences={showDifferences}
        dataSetChangeHandler={dataSetChangeHandler}
        roundPeriodChangeHandler={roundPeriodChangeHandler}
        showTimeChangeHandler={showTimeChangeHandler}
        showDifferencesChangeHandler={showDifferencesChangeHandler}
      />
      <Level>
        <Level.Item>
          <ResponsiveContainer width="100%" height={550}>
            <LineChart
              data={convertedMetadata}
              margin={{
                top: 0,
                right: 0,
                left: 0,
                bottom: 0,
              }}
            >
              <CartesianGrid strokeDasharray="3 3" />
              <XAxis
                dataKey="timestamp"
                type="number"
                domain={['dataMin', 'dataMax']}
                tickFormatter={showTime ? dateTimeFormatter : dateFormatter}
              />
              <YAxis
                allowDecimals={false}
                type="number"
                domain={['dataMin', 'dataMax']}
              >
                <Label value={dataSetTitlesMap[dataSet]} angle={-90} offset={5} position="insideLeft" />
              </YAxis>
              <Tooltip
                formatter={tooltipFormatter(titlesMap, differencesMap)}
                labelFormatter={showTime ? dateTimeFormatter : dateFormatter}
              />
              <Legend formatter={legendFormatter(titlesMap)} />
              {content}
            </LineChart>
          </ResponsiveContainer>
        </Level.Item>
      </Level>
    </>
  );
}

Chart.propTypes = {
  dataSetTitlesMap: PropTypes.instanceOf(Map).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  metadata: PropTypes.object.isRequired,
  titlesMap: PropTypes.instanceOf(Map).isRequired,
  dataSet: PropTypes.number.isRequired,
  roundPeriod: PropTypes.number.isRequired,
  showTime: PropTypes.bool.isRequired,
  showDifferences: PropTypes.bool.isRequired,
  dataSetChangeHandler: PropTypes.func.isRequired,
  roundPeriodChangeHandler: PropTypes.func.isRequired,
  showTimeChangeHandler: PropTypes.func.isRequired,
  showDifferencesChangeHandler: PropTypes.func.isRequired,
};
