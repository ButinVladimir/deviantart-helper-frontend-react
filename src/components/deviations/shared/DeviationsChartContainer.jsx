import { connect } from 'react-redux';
import Chart from '../../chart/Chart';
import * as deviationChart from '../../../redux/action-creators/deviations/chart/change-form-field-values';
import { dataSetOptions } from '../../../consts/data-sets';

/**
 * @description
 * Maps Redux state to props.
 *
 * @param {State} state - The Redux state.
 * @returns {Object} Props.
 */
const mapStateToProps = state => ({
  dataSetTitlesMap: new Map(dataSetOptions),
  dataSet: state.deviations.chart.dataSet,
  roundPeriod: state.deviations.chart.roundPeriod,
  showTime: state.deviations.chart.showTime,
  showDifferences: state.deviations.chart.showDifferences,
});

/**
 * @description
 * Maps Redux dispatch to props.
 *
 * @param {Function} dispatch - The Redux dispatch.
 * @returns {Object} Props.
 */
const mapDispatchToProps = dispatch => ({
  dataSetChangeHandler:
    e => dispatch(deviationChart.changeDataSetActionCreator(e.target.value)),
  roundPeriodChangeHandler:
    e => dispatch(deviationChart.changeRoundPeriodActionCreator(e.target.value)),
  showTimeChangeHandler:
    e => dispatch(deviationChart.changeShowTimeActionCreator(e.target.checked)),
  showDifferencesChangeHandler:
    e => dispatch(deviationChart.changeShowDifferencesActionCreator(e.target.checked)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Chart);
