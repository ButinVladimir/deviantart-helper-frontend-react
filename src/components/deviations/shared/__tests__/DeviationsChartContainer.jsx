import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import createDefaultStore from '../../../../redux/states/state';
import DeviationsChartContainer from '../DeviationsChartContainer';
import { DATA_SET_DOWNLOADS } from '../../../../consts/data-sets';
import { ROUND_PERIOD_4_HOURS } from '../../../../consts/round-periods';
import * as deviationsChart from '../../../../redux/action-creators/deviations/chart/change-form-field-values';

jest.mock(
  '../../../../redux/action-creators/deviations/chart/change-form-field-values',
  () => ({
    __esModule: true,
    changeDataSetActionCreator: jest.fn(() => ({ type: 'CHANGE_DATA_SET' })),
    changeRoundPeriodActionCreator: jest.fn(() => ({ type: 'CHANGE_ROUND_PERIOD' })),
    changeShowTimeActionCreator: jest.fn(() => ({ type: 'CHANGE_SHOW_TIME' })),
    changeShowDifferencesActionCreator: jest.fn(() => ({ type: 'CHANGE_SHOW_DIFFERENCES' })),
  }),
);

describe('DeviationsChartContainer', () => {
  const createStore = () => {
    const mockStore = configureStore([]);

    return mockStore({
      ...createDefaultStore(),
      deviations: {
        chart: {
          dataSet: DATA_SET_DOWNLOADS,
          roundPeriod: ROUND_PERIOD_4_HOURS,
          showTime: true,
          showDifferences: true,
        },
      },
    });
  };

  const metadata = { 1: [[1, 2, 3], [4, 5, 6]] };
  const titlesMap = new Map();
  titlesMap.set(1, 'Test entity');

  const props = { metadata, titlesMap };

  it('can be rendered correctly', () => {
    const store = createStore();
    const wrapper = mount(
      <Provider store={store}>
        <DeviationsChartContainer {...props} />
      </Provider>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('can handle changing data set', () => {
    const store = createStore();
    const wrapper = mount(
      <Provider store={store}>
        <DeviationsChartContainer {...props} />
      </Provider>,
    );

    wrapper.find('select#data-set').simulate('change');
    const actions = store.getActions();

    expect(deviationsChart.changeDataSetActionCreator).toHaveBeenCalled();
    expect(actions).toEqual([{ type: 'CHANGE_DATA_SET' }]);
  });

  it('can handle changing round period', () => {
    const store = createStore();
    const wrapper = mount(
      <Provider store={store}>
        <DeviationsChartContainer {...props} />
      </Provider>,
    );

    wrapper.find('select#round-period').simulate('change');
    const actions = store.getActions();

    expect(deviationsChart.changeRoundPeriodActionCreator).toHaveBeenCalled();
    expect(actions).toEqual([{ type: 'CHANGE_ROUND_PERIOD' }]);
  });

  it('can handle changing show time', () => {
    const store = createStore();
    const wrapper = mount(
      <Provider store={store}>
        <DeviationsChartContainer {...props} />
      </Provider>,
    );

    wrapper.find('input#show-time').simulate('change');
    const actions = store.getActions();

    expect(deviationsChart.changeShowTimeActionCreator).toHaveBeenCalled();
    expect(actions).toEqual([{ type: 'CHANGE_SHOW_TIME' }]);
  });

  it('can handle changing show differences', () => {
    const store = createStore();
    const wrapper = mount(
      <Provider store={store}>
        <DeviationsChartContainer {...props} />
      </Provider>,
    );

    wrapper.find('input#show-differences').simulate('change');
    const actions = store.getActions();

    expect(deviationsChart.changeShowDifferencesActionCreator).toHaveBeenCalled();
    expect(actions).toEqual([{ type: 'CHANGE_SHOW_DIFFERENCES' }]);
  });
});
