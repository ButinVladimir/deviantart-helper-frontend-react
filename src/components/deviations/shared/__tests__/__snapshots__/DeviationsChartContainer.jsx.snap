// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviationsChartContainer can be rendered correctly 1`] = `
Array [
  <div
    className="columns is-multiline"
    style={Object {}}
  >
    <div
      className="column"
      style={Object {}}
    >
      <div
        className="field"
        style={Object {}}
      >
        <label
          className="label"
          htmlFor={null}
          style={Object {}}
        >
          Data set
        </label>
        <div
          className="control"
          style={Object {}}
        >
          <div
            className="select"
            style={Object {}}
          >
            <select
              disabled={false}
              id="data-set"
              multiple={false}
              name="data-set"
              onChange={[Function]}
              readOnly={false}
              value={4}
            >
              <option
                key="1"
                value={1}
              >
                Views
              </option>
              <option
                key="2"
                value={2}
              >
                Favourites
              </option>
              <option
                key="3"
                value={3}
              >
                Comments
              </option>
              <option
                key="4"
                value={4}
              >
                Downloads
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
    <div
      className="column"
      style={Object {}}
    >
      <div
        className="field"
        style={Object {}}
      >
        <label
          className="label"
          htmlFor={null}
          style={Object {}}
        >
          Round period
        </label>
        <div
          className="control"
          style={Object {}}
        >
          <div
            className="select"
            style={Object {}}
          >
            <select
              disabled={false}
              id="round-period"
              multiple={false}
              name="round-period"
              onChange={[Function]}
              readOnly={false}
              value={14400000}
            >
              <option
                key="1"
                value={1}
              >
                No rounding
              </option>
              <option
                key="60000"
                value={60000}
              >
                1 minute
              </option>
              <option
                key="600000"
                value={600000}
              >
                10 minutes
              </option>
              <option
                key="3600000"
                value={3600000}
              >
                1 hour
              </option>
              <option
                key="14400000"
                value={14400000}
              >
                4 hours
              </option>
              <option
                key="43200000"
                value={43200000}
              >
                12 hours
              </option>
              <option
                key="86400000"
                value={86400000}
              >
                1 day
              </option>
              <option
                key="604800000"
                value={604800000}
              >
                7 days
              </option>
              <option
                key="864000000"
                value={864000000}
              >
                10 days
              </option>
              <option
                key="2592000000"
                value={2592000000}
              >
                30 days
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
    <div
      className="column"
      style={Object {}}
    >
      <div
        className="field"
        style={Object {}}
      >
        <label
          className="label"
          htmlFor={null}
          style={Object {}}
        >
          Show time
        </label>
        <div
          className="control"
          style={Object {}}
        >
          <label
            className="checkbox"
            disabled={false}
            style={Object {}}
          >
            <input
              checked={true}
              disabled={false}
              id="show-time"
              name="show-time"
              onChange={[Function]}
              type="checkbox"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
    <div
      className="column is-narrow"
      style={Object {}}
    >
      <div
        className="field"
        style={Object {}}
      >
        <label
          className="label"
          htmlFor={null}
          style={Object {}}
        >
          Show differences
        </label>
        <div
          className="control"
          style={Object {}}
        >
          <label
            className="checkbox"
            disabled={false}
            style={Object {}}
          >
            <input
              checked={false}
              disabled={false}
              id="show-differences"
              name="show-differences"
              onChange={[Function]}
              type="checkbox"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="level"
    style={Object {}}
  >
    <div
      className="level-item"
      style={Object {}}
    >
      <div
        className="recharts-responsive-container"
        style={
          Object {
            "height": 550,
            "maxHeight": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "none",
              "height": 0,
              "position": "absolute",
              "visibility": "hidden",
              "width": 0,
            }
          }
        />
      </div>
    </div>
  </div>,
]
`;
